kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,206 @@
+local buffer = require "string.buffer"
+local wasm = require "kong.runloop.wasm"
+local wasmx_shm
+
+
+local fmt = string.format
+local str_find = string.find
+local str_match = string.match
+local str_sub = string.sub
+local table_insert = table.insert
+local table_sort = table.sort
+local buf_new = buffer.new
+local ngx_say = ngx.say
+
+
+local _M = {}
```

Comment:
The first impression I had from seeing this file was that this looks pretty tightly coupled to the implementation details of ngx_wasm_module's metrics storage (well, and also to Prometheus, I suppose). Is it not feasible to put this code into `resty.wasmx.*` with a more simplified API that Kong gateway can use to register/inspect/collect metric [meta]data?

Reply:
I agree that the functions `parse_pw_key` and `serialize_metric` seem coupled with underlying implementation details.

As you noticed, `serialize_metric` is coupled with Prometheus serialization format, which can't really be avoided given the intent of the module.

`parse_pw_key`, on the other hand, is coupled with how labels are made part of the metric name by PW filter developers. This bit in fact seems to belong somewhere else as it'd have to be repeated when serializing to other formats (e.g. OpenTelemetry).

But I don't really think it belongs to `resty.wasmx.*`. Partially because `parse_pw_key` depends on filter metadata, but also because the concept of a label doesn't exist in ngx_wasmx_module, which represents a metric as a pair of a string and a data structure representing its value.

I would say having `parse_pw_key` logic part of this file isn't a big problem until we implement another form of serialization for Proxy-Wasm metrics.